src/car_backed_blockstore.rs

```diff
@@ -0,0 +1,330 @@
+//!  ┌───────────┼───┬──────────┤
+//!  │body length│cid│block data│
+//!  └───────────┴───┼──────────┤
+//!                  │◄────────►│
+//!                  │  =block data length
+//!      block data  │
+//!          offset ►│
+//! ```
+//!
+//! ## Block ordering
+//! > _... a filecoin-deterministic car-file is currently implementation-defined as containing all DAG-forming blocks in first-seen order, as a result of a depth-first DAG traversal starting from a single root._
+//!
+//! - [CAR documentation](https://ipld.io/specs/transport/car/carv1/#determinism)
+//!
+//! # Future work
+//! - [`fadvise`](https://linux.die.net/man/2/posix_fadvise)-based APIs to pre-fetch parts of the file, to improve random access performance.
+//! - Use an inner [`Blockstore`] for writes.
+//! - Support compressed snapshots.
+//!   Note that [`zstd`](https://github.com/facebook/zstd/blob/e4aeaebc201ba49fec50b087aeb15343c63712e5/doc/zstd_compression_format.md#zstandard-frames) archives are also composed of frames.
+//!   Snapshots typically comprise of a single frame, but that would require decompressing all preceding data, precluding random access.
+//!   So compressed snapshot support would compression per-frame, or maybe per block of frames.
+//! - Support multiple files by concatenating them.
+//! - Use safe arithmetic for all operations - a malicious frame shouldn't cause a crash.
+//! - Theoretically, [`CarBackedBlockstore`] should be clonable (or even [`Sync`]) with very low overhead, so that multiple threads could perform operations concurrently.
+
+use ahash::AHashMap;
+use cid::Cid;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_car::CarHeader;
+use parking_lot::Mutex;
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::io::{
+    self,
+    ErrorKind::{InvalidData, Other, UnexpectedEof, Unsupported},
+    Read, Seek, SeekFrom,
+};
+use tracing::{debug, trace};
+
+/// If you seek to `offset` (from the start of the file), and read `length` bytes,
+/// you should get data that corresponds to a [`Cid`] (but NOT the [`Cid`] itself).
+#[derive(Debug)]
+struct BlockDataLocation {
+    offset: u64,
+    length: u32,
+}
+
+/// See [module documentation](mod@self) for more.
+pub struct CarBackedBlockstore<ReaderT> {
+    // https://github.com/ChainSafe/forest/issues/3096
+    inner: Mutex<CarBackedBlockstoreInner<ReaderT>>,
+}
+
+impl<ReaderT> CarBackedBlockstore<ReaderT>
+where
+    ReaderT: Read + Seek,
+{
+    pub fn roots(&self) -> Vec<Cid> {
+        self.inner.lock().roots.clone()
+    }
+
+    /// To be correct:
+    /// - `reader` must read immutable data. e.g if it is a file, it should be [`flock`](https://linux.die.net/man/2/flock)ed.
+    ///   [`Blockstore`] API calls may panic if this is not upheld.
+    /// - `reader`'s buffer should have enough room for the [`CarHeader`] and any [`Cid`]s.
+    #[tracing::instrument(level = "debug", skip_all)]
+    pub fn new(mut reader: ReaderT) -> cid::Result<Self> {
```

Comment:
Would wrapping `reader` in a `BufReader` improve performance during the indexing?

Reply:
I've tested a few buffer sizes, on a 36G snapshot and a 2.6G snapshot

```console
$ du -h ...
36G     /home/aatif/chainsafe/snapshots/filecoin_full_calibnet_2023-04-07_450000.car
2.6G    /home/aatif/chainsafe/snapshots/forest_snapshot_calibnet_2023-06-29_height_690463.car
```

# Large snapshot
Note there's a low confidence here - n=1

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./target/release/examples/benchmark --mode buffer8k /home/aatif/chainsafe/snapshots/filecoin_full_calibnet_2023-04-07_450000.car` | 226.545 | 226.545 | 226.545 | 1.05 |
| `./target/release/examples/benchmark --mode buffer1k /home/aatif/chainsafe/snapshots/filecoin_full_calibnet_2023-04-07_450000.car` | 237.063 | 237.063 | 237.063 | 1.09 |
| `./target/release/examples/benchmark --mode buffer100 /home/aatif/chainsafe/snapshots/filecoin_full_calibnet_2023-04-07_450000.car` | 216.535 | 216.535 | 216.535 | 1.00 |
| `./target/release/examples/benchmark --mode unbuffered /home/aatif/chainsafe/snapshots/filecoin_full_calibnet_2023-04-07_450000.car` | 234.794 | 234.794 | 234.794 | 1.08 |

# Small snapshot
Probably not worth a lot, because the OS will probably keep a hot cache of the file on disk

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./target/release/examples/benchmark --mode buffer8k /home/aatif/chainsafe/snapshots/forest_snapshot_calibnet_2023-06-29_height_690463.car` | 3.229 ± 0.248 | 2.946 | 3.660 | 1.32 ± 0.13 |
| `./target/release/examples/benchmark --mode buffer1k /home/aatif/chainsafe/snapshots/forest_snapshot_calibnet_2023-06-29_height_690463.car` | 2.443 ± 0.148 | 2.276 | 2.668 | 1.00 |
| `./target/release/examples/benchmark --mode buffer100 /home/aatif/chainsafe/snapshots/forest_snapshot_calibnet_2023-06-29_height_690463.car` | 2.562 ± 0.305 | 2.149 | 3.043 | 1.05 ± 0.14 |
| `./target/release/examples/benchmark --mode unbuffered /home/aatif/chainsafe/snapshots/forest_snapshot_calibnet_2023-06-29_height_690463.car` | 5.788 ± 0.192 | 5.573 | 6.105 | 2.37 ± 0.16 |


Note that `std::io::Seek::seek`-ing on a `std::io::BufReader` always discards the buffer.

So _maybe_ a small buffer is good for our use-case here? I'd say it's not clear, and I don't want to spend too much time benchmarking. Gonna have a 100 byte buffer and call it at that, please do double check my benchmarks if you've got time: here's the code https://github.com/ChainSafe/forest/commit/7a20c4c9e4b7b583ec7d7593c823e0688e89e374

Reply:
Unabridged results over 10 runs:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `./target/release/examples/benchmark --mode buffer8k ../lotus/filecoin_full_calibnet_2023-04-07_450000.car` | 52.861 ± 0.587 | 52.156 | 54.317 | 1.02 ± 0.02 |
| `./target/release/examples/benchmark --mode buffer1k ../lotus/filecoin_full_calibnet_2023-04-07_450000.car` | 52.028 ± 0.687 | 51.455 | 53.880 | 1.00 |
| `./target/release/examples/benchmark --mode buffer100 ../lotus/filecoin_full_calibnet_2023-04-07_450000.car` | 53.525 ± 0.638 | 52.877 | 55.187 | 1.03 ± 0.02 |
| `./target/release/examples/benchmark --mode unbuffered ../lotus/filecoin_full_calibnet_2023-04-07_450000.car` | 85.451 ± 1.011 | 84.166 | 87.108 | 1.64 ± 0.03 |